[FEAT] 강의 생성 API 구현

JSeHoone opened the pull request:
## ❗️ 이슈 번호
#49 

## 📝 작업 내용
1. Lecture create request DTO에서 Validation을 적용하여 필수 응답 값을 설정하였습니다.
2. file의 정보를 추출하고 저장하기 위한 UploadFileInfoDTO를 추가하였습니다.
3. LectureServiceImpl에서 파일정보를 추출하는 메서드와 duration을 추출하는 메서드를 추가하였습니다.
4. video의 duration을 추출하기 위해 tika 라이브러리를 추가하였습니다.
5. 업로드 **파일의 형식은 `mp4`와 `pdf`만 가능**하도록 구현하였습니다.

## 💭 주의 사항
tika라는 라이브러리를 사용하더라도 모든 비디오의 시간을 추출하진 못했습니다.
이를 대체하기 위해서는 ffmpeg 라이브러리를 사용하는 것 같은데, 내장 java 라이브러리의 크기가 200MB까지 나온다고 합니다.
duration을 추출하기 위해서 해당 라이브러리를 사용한다는 것은 무리라고 생각이 됩니다.

그래서 duration 필드가 필요한가 의문이 들었습니다.

## 💡 리뷰 포인트
- duration 필드의 필요 여부를 논의하고 싶습니다.
- Service layer에 추가한 fileinfo를 추출한 부분에 문제가 없는지 확인 부탁드립니다 !

JSeHoone committed:
feat : Lecture 생성

- factory method를 사용하여 Lecture 객체를 생성하도록 구현

JSeHoone committed:
feat : course정보랑 user 정보를 가져오기

- section의 course가 lazy로 설정하여 해당 쿼리문을 작성하여 course 정보와 user 정보를 가져오도록 구현하였습니다.

JSeHoone committed:
feat : Apache tika 라이브러리 추가

- 동영상 시간을 추출하기 위한 의존성 추가

JSeHoone committed:
feat : Lecture create request, response DTO 추가

JSeHoone committed:
feat : 파일 정보를 추출 후 저장하기 위한 DTO

JSeHoone committed:
feat : Lecture Create API 추가

JSeHoone committed:
refactor : 세션이 없는 경우 404로 반환하도록 수정

dnya0 commented on src/main/java/com/example/projectlxp/lecture/service/impl/LectureServiceImpl.java:
역시 세훈님은 다르시군여.. 👍👍👍

그런데 `MultipartFile`과 관련된 부분은 `LectureService`가 갖고있기엔 너무 큰 책임인 것 같습니다.. 따로 file이나 resource와 같은 패키지 아래에 service로 분리하는 것을 추천드립니다!

JSeHoone replied:
유데미나 인프런에서는 Content라는 용어를 사용해서 하는 것 같아서 Lecture 패키지 내부에서 ContentService를 생성하겠습니다 !

감사합니다 :)

JSeHoone committed:
feat : ContentService 추가

- 업로드한 파일의 정보를 가공하는 역할을 합니다.
- 최종적으로는 파일을 파일 스토리지에 업로드하는 로직을 담당합니다.
- VIDEO 타입인 경우 영상 시간을 추출하는 `getVideoDuration` 메서드도 제공합니다.

JSeHoone committed:
refactor : ContentService로 책임 분리

- ContentService 의존성 주입하여 file정보를 가공하는 로직을 분리하였습니다.

dnya0 commented on src/main/java/com/example/projectlxp/content/service/ContentService.java:
이 Service를 다른 도메인 패키지로 분리하는 것이 좋아보입니다!
다른 도메인에서도 이미지와같은 파일을 불러오기 때문에 `MultipartFile`관련 로직을 다른 도메인으로 분리한다면 재사용성이 높아질 것 같네요! 지금은 동영상과 pdf에 대한 로직만 있지만 이후에 이미지를 추가할 수 있는거니까요

JSeHoone committed:
refactor : DTO를 record로 변경

- 기존 코드에서 public으로 설정해둬서, 변경이 가능한 문제가 있었음 !
- record를 사용해서 필드에 접근이 불가능하도록 수정

JSeHoone committed:
refactor : ContentService를 도메인 단위로 분리

JSeHoone committed:
refactor : UploadFileInfoDTO 리팩토링

- 해당 DTO의 위치를 content 도메인으로 옮겼습니다.
- DTO를 record로 변경하였습니다.

jjeongdong commented on src/main/java/com/example/projectlxp/lecture/service/impl/LectureServiceImpl.java:
```diff
@@ -0,0 +1,79 @@
+                                                "없는 세션입니다.", HttpStatus.NOT_FOUND));
+
+        // check who create this lecture
+        System.out.println(findSection.getCourse().getInstructor().getId());
```
불필요한 출력문 제거하면 졸을 것 같습니다.

jjeongdong commented on src/main/java/com/example/projectlxp/lecture/service/impl/LectureServiceImpl.java:
```diff
@@ -0,0 +1,79 @@
+        System.out.println(findSection.getCourse().getInstructor().getId());
+        if (!findSection.getCourse().getInstructor().getId().equals(userId)) {
+            throw new CustomBusinessException("강의 등록 권한이 없습니다.");
+        }
```
이것도 동일하게 따로 분리하면 졸을 것 같습니다.

jjeongdong commented on build.gradle:
```diff
@@ -47,6 +47,10 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-security'
     testImplementation 'org.springframework.security:spring-security-test'
 
+    // Apache Tika
+    implementation 'org.apache.tika:tika-core:2.9.2'
+    implementation 'org.apache.tika:tika-parsers-standard-package:2.9.2'
+
```
어떤 의존성인지 궁금합니다.

JSeHoone replied:
동영상의 duration을 계산하기 위해서 넣은 라이브러리입니다 !

근데, 동영상마다 추출이 되는 것도 있고, 안되는 것도 있어서 duration 필드에 대한 의문이 있긴 합니다 !..

JSeHoone committed:
refactor : Lecture 권한 체크하는 로직 분리

- Lecture가 속한 Section을 만든 사람과 다른 user가 lecture를 생성할 때 관리 권한 오류를 내뱉습니다.
